sdk/messaging/azservicebus/receiver.go

```diff
@@ -453,6 +485,26 @@ func (r *Receiver) receiveMessagesImpl(ctx context.Context, maxMessages int, opt
 	return receivedMessages, nil
 }
 
+// receiveFromCache gets any messages that were retrieved after the Receiver was closed
+// It returns an empty slice when the cache has been exhausted.
+func (r *Receiver) receiveFromCache(maxMessages int) []*ReceivedMessage {
+	cachedMessages := r.cachedMessages.Load()
+	if cachedMessages == nil || len(*cachedMessages) == 0 {
+		return nil
+	}
+
+	n := min(len(*cachedMessages), maxMessages)
+
+	receivedMessages := make([]*ReceivedMessage, n)
+
+	for i := range n {
+		receivedMessages[i] = newReceivedMessage((*cachedMessages)[i], nil)
+	}
+
+	(*cachedMessages) = (*cachedMessages)[n:]
```

Comment:
Only one goroutine can ever execute this statement?

Reply:
Yes - there's a mutex in ReceiveMessages(), which is driving all of this.